### ISSUE TITLE
Using list2padded for sequences

### ISSUE THREAD

**Issue by naveenjafer** [explosion/thinc#301]
I have referred to https://github.com/explosion/thinc/blob/master/examples/03_pos_tagger_basic_cnn.ipynb to get a better understanding of the thinc layers. The model is as follows in the example  
```
model = strings2arrays() >> with_array(
        HashEmbed(nO=width, nV=vector_width, column=0)
        >> expand_window(window_size=1)
        >> ReLu(nO=width, nI=width * 3)
        >> ReLu(nO=width, nI=width)
        >> Softmax(nO=nr_classes, nI=width)
    )
```
Can someone please explain what a string2arrays does? The documentation says that it takes a sequence of sequence of string and produces a list[int2D].  

The input X_train is something like [["this","is","awesome"],["thinc","is","cool"]]. What does the strings2arrays transform this example to? I am unable to wrap my head around what exactly strings2arrays does and how it transforms the input from a List[List] (2D lists) to List[Int2D] technically (3D lists/sequence).

**Comment by honnibal** [comment]
You're right that this is a bit confusing. I just had a look at the source, and the `strings2arrays` is inserting an extra dimension: https://github.com/explosion/thinc/blob/master/thinc/layers/strings2arrays.py

During development we had the `Embed` and `HashEmbed` layers expecting only 2d inputs, which didn't make that much sense. I coded the `strings2array` accordingly, making it return a 1-column array instead of the more obvious `ints1d`.

I think this should be changed: `strings2arrays` should return only `ints1d`. We'll have to update the examples accordingly.

**Comment by naveenjafer** [comment]
Oh that makes sense! I do have another continuing query. LSTM's need a padded sequence as input. In order to satisfy that, I assume the list2padded() layer should follow the string2array layer( Takes as input List[Array2D])? But when I do that, the execution of the model simply gets stuck without any progress. And since you are planning on reverting strings2array to a list[ints1d] would'nt the list2padded also need a change ( Assuming that these 2 layers need to be chained to feed it to with_array and subsequently the LSTM layer inside it. 

I was working on something like this, but it only gets stuck.
model = strings2arrays() >> list2padded() >> with_array(
        HashEmbed(nO=width, nV=vector_width, column=0)
        >> LSTM(nO=100, nI=width)
        >> Softmax(nO=nr_classes, nI=width)
    )

**Comment by honnibal** [comment]
@naveenjafer Ah you should use the `PyTorchLSTM` layer currently. I'm nearly done fixing the "native" LSTM implementation, but the one that's there currently is a draft and isn't very reliable.

In your model there, you have the `LSTM` inside the `with_array` block, which is not correct --- you need the `Padded` format to be input for the LSTM. I think the type-checking should highlight a problem with your architecture (although the errors are sometimes hard to read).

There are a couple of ways you could do the transformations. One is:

```
model = (
    strings2arrays()
    >> list2padded()
    >> with_array(HashEmbed(nO=width, nV=vector_width, column=0))
    >> PyTorchLSTM(nO=100, nI=width)
    >> with_array(Softmax(nO=nr_classes, nI=width))
    >> padded2list()
)
```

A couple of people have found the sequence formats confusing, so I'm thinking of ways to make it a bit simpler. For now, a good general strategy is to keep the data in a `List[Array]` format and use the `with_` transforms around blocks of your network. This is an easy way to get everything working, and then you can think about refactoring the transforms slightly if there's a more efficient way to do things.

For instance, here's how you could have your network: 

```
model = (
    strings2arrays()
    >> with_array(HashEmbed(nO=width, nV=vector_width, column=0))
    >> with_padded(PyTorchLSTM(nO=100, nI=width))
    >> with_array(Softmax(nO=nr_classes, nI=width))
)
```

**Comment by honnibal** [comment]
Pretty sure I've also found a bug in the `list2padded` method. Should have a fix shortly.

**Comment by naveenjafer** [comment]
@honnibal thank you so much for the update and the suggestion on how the layers need to be wrapped . 

I also see that the LSTM implementation does not have dropout implemented yet, would you be taking that up too in the coming stabilization? I would be able to give it a try in case no one has picked it up yet.

**Comment by honnibal** [comment]
I've got the CPU implementation mostly finished now, including weight-based dropout: https://github.com/explosion/thinc/pull/304

I didn't do activation dropout as I heard that's worse than doing the weight-based one. I need to double-check Smerity's AWD-LSTM paper though. If you want to check out that branch and try things out that would be very helpful. I should be able to get a CUDA implementation finished in the next week or two.

And yes, these are definitely things we want to be running well before the final 8.0.0 release. We also want a decent CNN implementation, suitable for image tasks, and multi-head attention.

**Comment by naveenjafer** [comment]
@honnibal So I ran a lil experiment wherein I ran the same model and dataset with PyTorchLSTM and LSTM. The Pytorch version works perfectly, it starts with accuracy of around 69 percent and improves with each iteration. The LSTM however starts with 15 % and hovers there for as long as I run it. I believe you might be aware of this already since you mentioned that it is not very reliable at this point, but 15 percent is probably way too low and is an indication of some small bu critical underlying issue?

**Comment by naveenjafer** [comment]
@honnibal 


> And yes, these are definitely things we want to be running well before the final 8.0.0 release. We also want a decent CNN implementation, suitable for image tasks, and multi-head attention.

I am interested in spending some time in implementing multi-head attention for sure, but would prefer starting with some slightly less complex layers to warm up. Is there any other pending layer that you have in mind that could be a good candidate?

**Comment by naveenjafer** [comment]
@honnibal and when you mention Multi head attention, are you referring to the smallest MHA unit with Q,K,V vectors? I initially misunderstood it as an implementation of the entire Attention is all you need transformers setup end to end as what you mean by an MHA layer.

**Comment by honnibal** [comment]
@naveenjafer Sorry I didn't see this before.

Yes I did mean just the multi-head attention layer, which wouldn't have any state associated to it. The transformer would then be built out of it and several other pieces. Anyway, I think the `list2padded` issue here should be fixed. so I'll close this issue as old.
